test/test_model.py

```diff
@@ -3391,17 +3395,44 @@ def test_opened_ports_warnings(self):
         self.assertIsInstance(ports_set, set)
         ports = sorted(ports_set, key=lambda p: (p.protocol, p.port))
         self.assertEqual(len(ports), 2)
-        self.assertIsInstance(ports[0], ops.OpenedPort)
+        self.assertIsInstance(ports[0], ops.Port)
         self.assertEqual(ports[0].protocol, 'tcp')
         self.assertEqual(ports[0].port, 8080)
-        self.assertIsInstance(ports[1], ops.OpenedPort)
+        self.assertIsInstance(ports[1], ops.Port)
         self.assertEqual(ports[1].protocol, 'udp')
         self.assertEqual(ports[1].port, 1000)
 
         self.assertEqual(fake_script_calls(self, clear=True), [
             ['opened-ports', ''],
         ])
 
+    def test_set_ports(self):
+        fake_script(self, 'open-port', 'exit 0')
+        fake_script(self, 'close-port', 'exit 0')
+        fake_script(self, 'opened-ports', 'exit 0')
+        self.unit.set_ports(8000, 8025)
+        calls = fake_script_calls(self, clear=True)
+        self.assertEqual(calls.pop(0), ['opened-ports', ''])
+        # We make no guarantee on the order the ports are opened.
+        calls.sort()
+        self.assertEqual(calls, [
+            ['open-port', '8000/tcp'],
+            ['open-port', '8025/tcp'],
+        ])
+        fake_script(self, 'opened-ports', 'echo 8025/tcp')
+        self.unit.set_ports(ops.Port('udp', 8022))
+        self.assertEqual(fake_script_calls(self, clear=True), [
+            ['opened-ports', ''],
+            ['close-port', '8025/tcp'],
+            ['open-port', '8022/udp'],
+        ])
+        fake_script(self, 'opened-ports', 'echo 8022/udp')
+        self.unit.set_ports()
+        self.assertEqual(fake_script_calls(self, clear=True), [
+            ['opened-ports', ''],
+            ['close-port', '8022/udp'],
+        ])
+
```

Comment:
So you have tested that you can open ports that haven't been declare,d and that you can replace existing ports with a new set of ports,and that you can close all ports, 

but you haven't tested that if a port is open and you call set_ports it is a no-op
or if there is a set of open ports that set_ports doesn't touch the ones that are the same. (eg, currently open are a,b and you ask for b,c should trigger a close of a, nothing for b, and an open of c)